guides/keras_cv/cut_mix_mix_up_and_rand_augment.py

```diff
@@ -0,0 +1,395 @@
+"""
+
+"""
+## RandAugment
+"""
+
+"""
+[RandAugment](https://arxiv.org/abs/1909.13719)
+has been shown to provide improved image
+classification results across numerous datasets.
+It performs a standard set of augmentations on an image.
+
+To use RandAugment in KerasCV, you need to provide a few values:
+
+- `value_range` describes the range of values covered in your images
+- `magnitude` is a value between 0 and 1, describing the strength of the perturbations
+applied
+- `augmentations_per_image` is an integer telling the layer how many augmentations to apply to each
+individual image
+- (Optional) `magnitude_stddev` allows `magnitude` to be randomly sampled
+from a distribution with a standard deviation of `magnitude_stddev`
+- (Optional) `rate` indicates the probability to apply the augmentation
+applied at each layer.
+
+You can read more about these
+parameters in the
+[`RandAugment` API documentation](/api/keras_cv/layers/rand_augment).
+
+Let's use KerasCV's RandAugment implementation.
+"""
+
+rand_augment = keras_cv.layers.RandAugment(
+    value_range=(0, 255),
+    augmentations_per_image=3,
+    magnitude=0.3,
+    magnitude_stddev=0.2,
+    rate=0.5,
+)
+
+
+def apply_rand_augment(inputs):
+    inputs["images"] = rand_augment(inputs["images"])
+    return inputs
+
+
+train_dataset = load_dataset().map(apply_rand_augment, num_parallel_calls=AUTOTUNE)
+
+"""
+Finally, let's inspect some of the results:
+"""
+
+visualize_dataset(train_dataset, title="After RandAugment")
+
+"""
+Try tweaking the magnitude settings to see a wider variety of results.
+"""
+
+"""
+## CutMix and MixUp: generate high-quality inter-class examples
+
+
+`CutMix` and `MixUp` allow us to produce inter-class examples. `CutMix` randomly cuts out
+portions of one image and places them over another, and `MixUp` interpolates the pixel
+values between two images. Both of these prevent the model from overfitting the
+training distribution and improve the likelihood that the model can generalize to out of
+distribution examples. Additionally, `CutMix` prevents your model from over-relying on
+any particular feature to perform its classifications. You can read more about these
+techniques in their respective papers:
+
+- [CutMix: Train Strong Classifiers](https://arxiv.org/abs/1905.04899)
+- [MixUp: Beyond Empirical Risk Minimization](https://arxiv.org/abs/1710.09412)
+
+In this example, we will use `CutMix` and `MixUp` independently in a manually created
+preprocessing pipeline. In most state of the art pipelines images are randomly
+augmented by either `CutMix`, `MixUp`, or neither. The function below implements this
+in an equal 1/3 split.
+
+Note that our `cut_mix_and_mix_up` function is annotated with a `tf.function` to ensure
+optimal performance.
+"""
+cut_mix = keras_cv.layers.CutMix()
+mix_up = keras_cv.layers.MixUp()
+
+
+def cut_mix_and_mix_up(samples):
+    samples = cut_mix(samples, training=True)
+    samples = mix_up(samples, training=True)
```

Comment:
Is it good practice to call those at the same time? That would be really heavy augmentation.

Reply:
Or, we can combine both here. Like

```python
def cut_mix_and_mix_up(samples):
    cutmix_samples = cut_mix(samples, training=True)
    mixup_samples = mix_up(samples, training=True)

    samples['images'] = tf.concat([cutmix_samples['images'], 
                                     mixup_samples['images']], axis=0)
    samples['labels'] = tf.concat([cutmix_samples['labels'], 
                                    mixup_samples['labels']], axis=0)

    return samples
```